File: leetcode/python/0008_string_to_integer.py
# https://leetcode.com/problems/string-to-integer-atoi/
# O(n) space complexity
# O(1) time complexity
def string_to_integer(s: str) -> int:
    MAX_NUM = 2 ** 31 - 1
    MIN_NUM = -2 ** 31

    total = 0
    index = 0
    sign = 1

    while index < len(s) and s[index] == ' ':
        index += 1

    if index < len(s) and (s[index] == '-' or s[index] == '+'):
        sign = -1 if s[index] == '-' else 1
        index += 1

    while index < len(s) and s[index].isdigit():
        digit = ord(s[index]) - ord('0')
        if total > (MAX_NUM - digit) / 10:
            return MAX_NUM if sign > 0 else MIN_NUM
        total = total * 10 + digit
        index += 1

    return total * sign

File: leetcode/java/0008_string_to_integer.java
// https://leetcode.com/problems/string-to-integer-atoi/
// O(n) time complexity
// O(1) space complexity
class Solution_0008 {
  public int myAtoi(String s) {
    if (s == null || s.length() == 0) return 0;
    int index = 0, sign = 1, result = 0;
    
    while (index < s.length() && s.charAt(index) == ' ') {
      index++;
    }
    
    if (index < s.length() && (s.charAt(index) == '-' || s.charAt(index) == '+')) {
      sign = s.charAt(index) == '-' ? -1 : 1;
      index++;
    }
    
    while (index < s.length() && Character.isDigit(s.charAt(index))) {
      int digit = s.charAt(index++) - '0';
      if (result > (Integer.MAX_VALUE - digit) / 10) {
        return sign == -1 ? Integer.MIN_VALUE : Integer.MAX_VALUE;
      }
      result = result * 10 + digit;
    }
    
    return result * sign;
  }
}
